File: zypher-scan-api/api/routes/route.py
from fastapi import FastAPI, APIRouter, HTTPException
from models.todos import Todo
from config.database import collection_name
from schema.schemas import list_serial
from bson import ObjectId

router = APIRouter(
    prefix='/todos',
    tags=['routines']
)

@router.get("/")
async def get_todos():
    try:
        todos = list_serial(collection_name.find())
        return todos
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error fetching todos: {str(e)}")


@router.post("/")
async def create_todo(todo: Todo):
    try:
        result = collection_name.insert_one(dict(todo))
        if not result.inserted_id:
            raise HTTPException(status_code=500, detail="Failed to create todo")
        return {"id": str(result.inserted_id), "message": "Todo created successfully"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error creating todo: {str(e)}")


@router.put("/{id}")
async def update_todo(id: str, todo: Todo):
    try:
        result = collection_name.find_one_and_update(
            {"_id": ObjectId(id)},
            {"$set": dict(todo)},
            return_document=True
        )
        if not result:
            raise HTTPException(status_code=404, detail="Todo not found")
        return {"message": "Todo updated successfully"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error updating todo: {str(e)}")


@router.delete("/{id}")
async def delete_todo(id: str):
    try:
        result = collection_name.find_one_and_delete({"_id": ObjectId(id)})
        if not result:
            raise HTTPException(status_code=404, detail="Todo not found")
        return {"message": "Todo deleted successfully"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error deleting todo: {str(e)}")

File: zypher-app/src/app/api/customRuleMetadata/route.js
import { NextResponse } from "next/server";
import CustomRuleMetadata from '../../../models/CustomRuleMetadata';
import connectDB from "@/utils/db";

export async function POST(req) {
  await connectDB();
  try {
    const body = await req.json();
    const rule = await CustomRuleMetadata.create(body);
    return NextResponse.json(rule, { status: 201 });
  } catch (err) {
    return NextResponse.json({ error: err.message }, { status: 400 });
  }
}

export async function GET() {
  await connectDB();
  try {
    const rules = await CustomRuleMetadata.find();
    return NextResponse.json(rules, { status: 200 });
  } catch (err) {
    return NextResponse.json({ error: err.message }, { status: 500 });
  }
}

export async function PUT(req) {
  await connectDB();
  try {
    const body = await req.json();
    const { rule_id } = body;
    const rule = await CustomRuleMetadata.findOneAndUpdate({ rule_id }, body, { new: true });
    if (!rule) return NextResponse.json({ error: 'Rule not found' }, { status: 404 });
    return NextResponse.json(rule, { status: 200 });
  } catch (err) {
    return NextResponse.json({ error: err.message }, { status: 400 });
  }
}

export async function DELETE(req) {
  await connectDB();
  try {
    const body = await req.json();
    const { rule_id } = body;
    const result = await CustomRuleMetadata.deleteOne({ rule_id });
    return NextResponse.json(result, { status: 200 });
  } catch (err) {
    return NextResponse.json({ error: err.message }, { status: 400 });
  }
}
